### PR TITLE
Twisted callback

### PR DESCRIPTION
**Pull request by whirm**
Do not merge yet, I'm extending this PR to include the complete callback usage removal from dispersy.

### CODE REVIEW

**Comment by whirm** [comment]
retest this please

File: bootstrap.py [hunk 1]
```diff
@@ -137,48 +141,22 @@ def resolve(self, func, timeout=60.0, blocking=False):
...
         assert timeout > 0.0, timeout
-        assert isinstance(blocking, bool), type(blocking)
 
         if self.are_resolved:
-            self._callback.register(func, (True,))
-
+            reactor.callLater(0, func, True)
         else:
+            # TODO(emilon): What is this used for? Can I remove it?
             self._thread_counter += 1
```
Comment (NielsZeilemaker) [inline bootstrap.py]: Was used to create a unique name for each callback resolve instance.

File: tests/test_callback.py [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
+
+from unittest import skip
+
 from .dispersytestclass import DispersyTestFunc, call_on_mm_thread
 
 
 class TestCallback(DispersyTestFunc):
 
-    @call_on_mm_thread
+    #@call_on_mm_thread
```
Comment (NielsZeilemaker) [inline tests/test_callback.py]: These need to be on the mm thread because of the yield. Now they are bascially not running.

File: dispersy.py [hunk 3]
```diff
@@ -319,16 +320,14 @@ def retry_until_success():
...
             # first attempt will block for at most TIMEOUT seconds
             logger.debug("resolving bootstrap addresses (%.1s timeout)", timeout)
             # give low priority to ensure that on_results is called before the call returns
-            self._callback.call(bootstrap.resolve, kargs=dict(func=on_results, timeout=timeout, blocking=True), priority= -128)
 
-            if not bootstrap.are_resolved:
-                # unable to resolve all... retry until successful
-                self._callback.register(retry_until_success, delay=300.0)
+            # TODO(emilon): Since we moved to twisted, this call will not block, but all this code should be refactored soon anyway
+            self._callback.register(bootstrap.resolve, kargs=dict(func=on_results, timeout=timeout))
```
Comment (NielsZeilemaker) [inline dispersy.py]: It has to block, i'm not sure how the search community fast walker thingy will respond to not having any bootstrap candidates.

File: tests/test_dynamicsettings.py [hunk 4]
```diff
@@ -22,6 +23,7 @@ def test_default_resolution(self):
 
         # NODE creates a message (should allow, because the default policy is PublicResolution)
         message = self._mm.create_dynamic_resolution_text("Message #%d" % 10, policy.implement(), 10)
+        #self._mm.give_message(message, self._mm)
```
Comment (NielsZeilemaker) [inline tests/test_dynamicsettings.py]: I guess these commented lines can be removed

**Comment by NielsZeilemaker** [comment]
It's looking great, I have just one remark. Maybe remove the Callback object alltogether, and merge some of its methods to the twistedcallback class.
We're never going back are we :smile:

**Comment by whirm** [comment]
@NielsZeilemaker yeah, I planned to do that when the bootstraping was twistified and the tracker working (doing the later now)

File: dispersy.py [hunk 5]
```diff
@@ -301,36 +297,23 @@ def on_results(success):
...
-
-            if not bootstrap.are_resolved:
-                # unable to resolve all... retry until successful
-                self._callback.register(retry_until_success, delay=300.0)
+            # first attempt will block for at most TIMEOUT seconds
+            result = yield bootstrap.resolve(timeout=timeout).addCallback(on_results)
+            now = False
 
-        return bootstrap.are_resolved
+        bootstrap.resolve_until_success(now)
```
Comment (NielsZeilemaker) [inline dispersy.py]: now=now right?

File: tests/test_destroycommunity.py [hunk 6]
```diff
@@ -25,9 +32,11 @@ def test_hard_kill_without_permission(self):
         node, other = self.create_nodes(2)
         node.send_identity(other)
 
-        message = node.create_full_sync_text("Should be remove" , 42)
+        message = node.create_full_sync_text("Should not be removed", 42)
         node.give_message(message, node)
 
+        node.process_packets()
```
Comment (NielsZeilemaker) [inline tests/test_destroycommunity.py]: shouldn't be neccesary

File: dispersy.py (hunk 5 above)
Comment (whirm) [inline dispersy.py]: oops, yes, thanks

File: tests/test_destroycommunity.py (hunk 6 above)
Comment (whirm) [inline tests/test_destroycommunity.py]: let's try :D

File: callback.py [hunk 7]
```diff
@@ -803,3 +814,250 @@ def loop(self):
...
+                if timeout:
+                    logger.warning("Ignoring timeout as we are being called from the twisted thread")
+                if id_:
+                    logger.warning("Ignoring id_ as we are being called from the twisted thread")
+                return call(*args, **kargs)
+            else:
+                id_ = self.generate_id(id_)
+
+                # TODO(emilon): once register() is gone, we could use this instead of all this mess:
+                # result = blockingCallFromThread(reactor, call, *args, **kargs)
```
Comment (NielsZeilemaker) [inline callback.py]: we can't really as blockingcall doesnt support a timeout

Comment (whirm) [inline callback.py]: hm, right

**Comment by LipuFei** [comment]
@whirm Could you do a rebase on devel? Then you'll be able to see clearly what I've done when I make a pull request to your branch. Thanks.

**Comment by whirm** [comment]
@LipuFei Sure, il'l later

**Comment by whirm** [comment]
@LipuFei I rebased a while ago. Right now I'm forking off devel's HEAD

File: community.py [hunk 8]
```diff
@@ -378,9 +388,12 @@ def on_dispersy_identity(message):
                 assert message.authentication.member.mid == self._master_member.mid
                 self._master_member = message.authentication.member
                 assert self._master_member.public_key
+                if self._master_member_dc.active():
+                    self._master_member_dc.cancel()
+                self._pending_tasks.remove(self._master_member_dc)
 
-        delay = 2.0
-        while not self._master_member.public_key:
+        def fetch_master_member_identity(delay):
```
Comment (NielsZeilemaker) [inline community.py]: Can we convert this one to a looping call?

File: dispersy.py [hunk 9]
```diff
@@ -169,17 +165,11 @@ def __init__(self, callback, endpoint, working_directory, database_filename=u"di
...
-                start = time()
-                try:
-                    while True:
-                        yield float(60 * 60)
-                        scanner.dump_all_objects("memory-%d.out" % (time() - start))
-                except GeneratorExit:
-                    scanner.dump_all_objects("memory-%d-shutdown.out" % (time() - start))
-
-            self._callback.register(memory_dump)
+            start = time()
```
Comment (NielsZeilemaker) [inline dispersy.py]: This "logic" should be moved to the decorator class, and then lets add the @start_memory_dump at the constructor of dispersy

File: dispersy.py [hunk 10]
```diff
@@ -2316,135 +2284,134 @@ def start(self, timeout=10.0):
...
-        3. opens database
-        4. opens endpoint
+        1. resolve bootstrap candidates (done in parallel)
+        2. opens database
+        3. opens endpoint
         """
 
-        assert not self._callback.is_running, "Must be called before callback.start()"
+        # Disabled this assert as is_running now checks if the reactor is running
+        # assert not self._callback.is_running, "Must be called before callback.start()"
```
Comment (NielsZeilemaker) [inline dispersy.py]: These two lines can be removed

File: dispersy.py [hunk 11]
```diff
@@ -2316,135 +2284,134 @@ def start(self, timeout=10.0):
...
+            raise RuntimeError("Dispersy is already running")
 
         # start
         logger.info("starting the Dispersy core...")
         results = []
 
-        results.append((u"callback", self._callback.start()))
         assert all(isinstance(result, bool) for _, result in results), [type(result) for _, result in results]
-        self._callback.call(start, priority=512)
+        assert isInIOThread(), "Must be called from the callback thread"
```
Comment (NielsZeilemaker) [inline dispersy.py]: duplicate

File: endpoint.py [hunk 12]
```diff
@@ -354,7 +359,11 @@ def _loop(self):
 
                 finally:
                     if packets:
-                        logger.debug('%d came in, %d bytes in total', len(packets), sum(len(packet) for _, packet in packets))
+                        if logger and logger.debug:
```
Comment (NielsZeilemaker) [inline endpoint.py:362]: This doesn't make sense, i'd rather not have checks for missing globals.
Let's address this by converting all loggers to a self._logger instead

File: requestcache.py [hunk 13]
```diff
@@ -175,14 +154,14 @@ def add(self, cache):
         else:
             logger.debug("add %s", cache)
             self._identifiers[identifier] = cache
-            cache.callback_identifier = self._callback.register(self._on_timeout, (cache,), delay=cache.timeout_delay)
+            cache.timeout_delayed_call = reactor.callLater(cache.timeout_delay, self._on_timeout, cache)
```
Comment (NielsZeilemaker) [inline requestcache.py]: Maybe we should opt for a second dictionary instead, Dispersy doesnt really use property assignmet as you're doing here.

File: tests/debugcommunity/node.py [hunk 14]
```diff
@@ -118,14 +126,18 @@ def init_my_member(self, tunnel=False, store_identity=True):
...
-
             # add this node to candidate list of mm
             message = self.create_introduction_request(self._central_node.my_candidate, self.lan_address, self.wan_address, False, u"unknown", None, 1, 1)
-            self._central_node.give_message(message, self)
+            yield self._central_node.give_message(message, self)
 
+            # TODO(emilon): Make it so it is no longer necessary to do this
             # remove introduction responses from socket
-            self.receive_messages(names=[u'dispersy-introduction-response'])
+            for _ in xrange(5):
```
Comment (NielsZeilemaker) [inline tests/debugcommunity/node.py]: Can we move this into the receive_messages?

File: tests/test_batch.py [hunk 15]
```diff
@@ -69,6 +69,7 @@ def test_one_big_batch(self, length=1000):
 
     def test_many_small_batches(self, length=1000):
         """
+        Test many small batches of messages is processed correctly.
```
Comment (NielsZeilemaker) [inline tests/test_batch.py]: are

File: tests/test_bootstrap.py [hunk 16]
```diff
@@ -14,13 +15,15 @@
...
-class TestBootstrapServers(DispersyTestFunc):
 
+class TestBootstrapServers(DispersyTestFunc):
 
     def test_tracker(self):
         """
         Runs tracker.py and connects to it.
         """
+
+        Bootstrap.enable = True
```
Comment (NielsZeilemaker) [inline tests/test_bootstrap.py:27]: Is it set to false anywhere?

File: tests/test_dynamicsettings.py [hunk 17]
```diff
@@ -46,6 +47,8 @@ def test_change_resolution(self):
         node.give_message(message, self._mm)
         other.give_message(message, self._mm)
 
+        #other.process_packets()
```
Comment (NielsZeilemaker) [inline tests/test_dynamicsettings.py]: Remove this line

File: tests/test_overlay.py [hunk 18]
```diff
@@ -60,6 +68,40 @@ def take_step(self):
             def dispersy_enable_fast_candidate_walker(self):
                 return enable_fast_walker
 
+            @inlineCallbacks
```
Comment (NielsZeilemaker) [inline tests/test_overlay.py]: Merge bug, the dispersy_enable_fast_candidate_walker property enabled the fast walker

File: tests/test_bootstrap.py (hunk 16 above)
Comment (whirm) [inline tests/test_bootstrap.py:27]: on the test class module

File: endpoint.py (hunk 12 above)
Comment (whirm) [inline endpoint.py:362]: we are going to address this on a separate pull request (logger.py removal)

File: bootstrap.py [hunk 19]
```diff
@@ -125,60 +127,44 @@ def reset(self):
         with self._lock:
             self._candidates = dict((address, None) for address in self._candidates.iterkeys())
 
-    def resolve(self, func, timeout=60.0, blocking=False):
+    @inlineCallbacks
+    def resolve(self, timeout=60.0):
```
Comment (NielsZeilemaker) [inline bootstrap.py]: The timeout argument isn't used, i guess we should remove it

File: community.py [hunk 20]
```diff
@@ -28,19 +34,19 @@
...
 from .statistics import CommunityStatistics
 from .timeline import Timeline
 
 
-from collections import OrderedDict
-
-
 logger = get_logger(__name__)
 
+DOWNLOAD_MM_PK_INTERVAL = 300.0
```
Comment (NielsZeilemaker) [inline community.py]: Can you reduce the interval to 15.0

File: community.py [hunk 21]
```diff
@@ -187,17 +194,17 @@ def __init__(self, dispersy, master, my_member, attach=True):
...
+        self._last_sync_time = 0
+
         # Dispersy
         self._dispersy = dispersy
 
-        # _pending_callbacks contains all id's for registered calls that should be removed when the
-        # community is unloaded.  most of the time this contains all the generators that are being
-        # used by the community
-        self._pending_callbacks = []
+        # _pending_callbacks contains all pending calls that should be removed when the
```
Comment (NielsZeilemaker) [inline community.py]: update to pending_tasks

File: community.py [hunk 22]
```diff
@@ -378,27 +388,23 @@ def on_dispersy_identity(message):
                 assert message.authentication.member.mid == self._master_member.mid
                 self._master_member = message.authentication.member
                 assert self._master_member.public_key
+                self._pending_tasks.pop("download master member identity").stop()
```
Comment (NielsZeilemaker) [inline community.py]: maybe call cancel_pending_task instead

File: community.py [hunk 23]
```diff
@@ -378,27 +388,23 @@ def on_dispersy_identity(message):
...
+        try:
+            public_key, = self._dispersy.database.execute(u"SELECT public_key FROM member WHERE id = ?", (self._master_member.database_id,)).next()
+        except StopIteration:
+            pass
+        else:
+            if public_key:
+                logger.debug("%s found master member", self._cid.encode("HEX"))
+                self._master_member = self._dispersy.get_member(public_key=str(public_key))
+                assert self._master_member.public_key
+                self._pending_tasks.pop("download master member identity").stop()
```
Comment (NielsZeilemaker) [inline community.py]: same here

File: dispersy.py [hunk 24]
```diff
@@ -68,12 +73,11 @@
...
 
-from collections import OrderedDict
-
 logger = get_logger(__name__)
 
-from .callback import Callback
-from .endpoint import Endpoint
+FLUSH_DATABASE_INTERVAL = 60.0
+STATS_DETAILED_CANDIDATES_INTERVAL = 5.0
+MEMORY_DUMP_INTERVAL = float(60 * 60)
```
Comment (NielsZeilemaker) [inline dispersy.py]: This one should be moved to utils

File: dispersy.py (hunk 9 above)
Comment (NielsZeilemaker) [inline dispersy.py]: Its still here, can you move it to the utils thingy

File: dispersy.py [hunk 25]
```diff
@@ -300,36 +291,23 @@ def on_results(success):
...
-            # first attempt will block for at most TIMEOUT seconds
             logger.debug("resolving bootstrap addresses (%.1s timeout)", timeout)
-            # give low priority to ensure that on_results is called before the call returns
-            self._callback.call(bootstrap.resolve, kargs=dict(func=on_results, timeout=timeout, blocking=True), priority= -128)
-
-            if not bootstrap.are_resolved:
-                # unable to resolve all... retry until successful
-                self._callback.register(retry_until_success, delay=300.0)
+            # first attempt will block for at most TIMEOUT seconds
+            result = yield bootstrap.resolve(timeout=timeout).addCallback(on_results)
```
Comment (NielsZeilemaker) [inline dispersy.py]: Resolve doesn't support a timeout anymore, which is unfortunate

File: tests/debugcommunity/node.py [hunk 26]
```diff
@@ -189,6 +190,7 @@ def send_message(self, message, candidate):
 
         return self.send_packet(message.packet, candidate)
 
+    #@call_on_dispersy_thread
```
Comment (NielsZeilemaker) [inline tests/debugcommunity/node.py]: remove this line

File: tests/test_overlay.py (hunk 18 above)
Comment (NielsZeilemaker) [inline tests/test_overlay.py]: This still needs to be removed

File: util.py [hunk 27]
```diff
@@ -0,0 +1,55 @@
+# util.py ---
```
Comment (NielsZeilemaker) [inline util.py]: I know you like these headers, but lets remove it for now.
If we're going to add a license, all file need a similar header, or we can simply add a license file.
- can you put all other decorators in here as well?

**Comment by whirm** [comment]
I'm still running the tests locally, but everything looks fine so far.

File: tests/test_overlay.py [hunk 28]
```diff
@@ -56,9 +65,39 @@ def take_step(self):
                 for sleep in Community.take_step(self):
                     yield sleep
 
-            @property
-            def dispersy_enable_fast_candidate_walker(self):
-                return enable_fast_walker
+            @inlineCallbacks
```
Comment (NielsZeilemaker) [inline tests/test_overlay.py]: This one should be removed, and the property reinstated

**Pull request merged**